File: internal/hosts/hosts.go
package hosts

import (
	"bufio"
	"fmt"
	"os"
	"runtime"
	"strings"
)

const (
	beginMarker = "# BEGIN hostler managed block"
	endMarker   = "# END hostler managed block"
)

// GetHostsPath returns the path to the hosts file
func GetHostsPath() string {
	if runtime.GOOS == "windows" {
		return `C:\Windows\System32\drivers\etc\hosts`
	}
	return "/etc/hosts"
}

// HasDomain checks if a domain exists in the hosts file
func HasDomain(hostsPath, domain string) (bool, error) {
	file, err := os.Open(hostsPath)
	if err != nil {
		return false, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if strings.Contains(line, domain) && !strings.HasPrefix(strings.TrimSpace(line), "#") {
			return true, nil
		}
	}
	return false, scanner.Err()
}

// AddEntry adds a domain to the hosts file in the managed block
func AddEntry(hostsPath, domain string) error {
	entries, err := getManagedEntries(hostsPath)
	if err != nil {
		return err
	}

	// Check if already exists
	for _, e := range entries {
		if e == domain {
			return nil // Already exists
		}
	}

	entries = append(entries, domain)
	return writeManagedBlock(hostsPath, entries)
}

// RemoveEntry removes a domain from the hosts file managed block
func RemoveEntry(hostsPath, domain string) error {
	entries, err := getManagedEntries(hostsPath)
	if err != nil {
		return err
	}

	var newEntries []string
	for _, e := range entries {
		if e != domain {
			newEntries = append(newEntries, e)
		}
	}

	return writeManagedBlock(hostsPath, newEntries)
}

// GetManagedDomains returns all domains in the managed block
func GetManagedDomains(hostsPath string) ([]string, error) {
	return getManagedEntries(hostsPath)
}

func getManagedEntries(hostsPath string) ([]string, error) {
	var entries []string

	file, err := os.Open(hostsPath)
	if err != nil {
		return entries, err
	}
	defer file.Close()

	inBlock := false
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()

		if strings.TrimSpace(line) == beginMarker {
			inBlock = true
			continue
		}
		if strings.TrimSpace(line) == endMarker {
			inBlock = false
			continue
		}

		if inBlock {
			// Parse "127.0.0.1	domain.loc" format
			fields := strings.Fields(line)
			if len(fields) >= 2 && fields[0] == "127.0.0.1" {
				entries = append(entries, fields[1])
			}
		}
	}

	return entries, scanner.Err()
}

func writeManagedBlock(hostsPath string, entries []string) error {
	// Read existing content
	data, err := os.ReadFile(hostsPath)
	if err != nil {
		return err
	}

	content := string(data)

	// Remove existing managed block
	var lines []string
	inBlock := false
	for _, line := range strings.Split(content, "\n") {
		if strings.TrimSpace(line) == beginMarker {
			inBlock = true
			continue
		}
		if strings.TrimSpace(line) == endMarker {
			inBlock = false
			continue
		}
		if !inBlock {
			lines = append(lines, line)
		}
	}

	// Build new managed block
	var newBlock strings.Builder
	if len(entries) > 0 {
		newBlock.WriteString(beginMarker + "\n")
		for _, domain := range entries {
			newBlock.WriteString(fmt.Sprintf("127.0.0.1\t%s\n", domain))
		}
		newBlock.WriteString(endMarker + "\n")
	}

	// Combine and write
	result := strings.Join(lines, "\n")
	// Remove trailing newlines
	result = strings.TrimRight(result, "\n")
	// Add our block at the end
	if len(entries) > 0 {
		result = result + "\n" + newBlock.String()
	} else {
		result = result + "\n"
	}

	return os.WriteFile(hostsPath, []byte(result), 0644)
}

File: src/lib/hosts.ts
// Manages /etc/hosts entries inside a delimited block so hostler never touches
// lines it doesn't own.
import { readFile } from "node:fs/promises";
import { withLock, writeFileAtomicSync } from "./fsatomic.ts";

const BEGIN_MARKER = "# BEGIN hostler managed block";
const END_MARKER = "# END hostler managed block";

/** Path to the system hosts file for the current platform. */
export function getHostsPath(): string {
  if (process.platform === "win32") {
    return "C:\\Windows\\System32\\drivers\\etc\\hosts";
  }
  return "/etc/hosts";
}

/** True if `domain` appears on any active (non-comment) line in the hosts file. */
export async function hasDomain(hostsPath: string, domain: string): Promise<boolean> {
  const content = await readFile(hostsPath, "utf8");
  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    if (trimmed.startsWith("#")) continue;
    // Match the domain as a whole field, not a substring of another host.
    if (trimmed.split(/\s+/).includes(domain)) return true;
  }
  return false;
}

/**
 * True if `domain` appears on an active line OUTSIDE hostler's managed block.
 * Such an entry would shadow (or compete with) the one hostler manages — and
 * hostler must not silently claim to have "updated" a line it doesn't own.
 */
export async function hasUnmanagedDomain(hostsPath: string, domain: string): Promise<boolean> {
  const content = await readFile(hostsPath, "utf8").catch(() => "");
  const target = domain.toLowerCase(); // hostnames are case-insensitive
  let inBlock = false;
  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    if (trimmed === BEGIN_MARKER) {
      inBlock = true;
      continue;
    }
    if (trimmed === END_MARKER) {
      inBlock = false;
      continue;
    }
    if (inBlock || trimmed.startsWith("#") || trimmed === "") continue;
    const hostFields = trimmed.split(/\s+/).slice(1).map((f) => f.toLowerCase());
    if (hostFields.includes(target)) return true;
  }
  return false;
}

/** Returns the domains currently inside hostler's managed block. */
export async function getManagedDomains(hostsPath: string): Promise<string[]> {
  const content = await readFile(hostsPath, "utf8").catch(() => "");
  const entries: string[] = [];
  let inBlock = false;
  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    if (trimmed === BEGIN_MARKER) {
      inBlock = true;
      continue;
    }
    if (trimmed === END_MARKER) {
      inBlock = false;
      continue;
    }
    if (inBlock) {
      const fields = trimmed.split(/\s+/);
      if (fields.length >= 2 && fields[0] === "127.0.0.1") {
        entries.push(fields[1]!);
      }
    }
  }
  return entries;
}

/** Adds a domain to the managed block (no-op if already present). */
export async function addEntry(hostsPath: string, domain: string): Promise<void> {
  // Lock the whole read-modify-write so concurrent invocations can't clobber.
  await withLock("hosts", async () => {
    const entries = await getManagedDomains(hostsPath);
    if (entries.includes(domain)) return;
    entries.push(domain);
    await writeManagedBlock(hostsPath, entries);
  });
}

/** Removes a domain from the managed block. */
export async function removeEntry(hostsPath: string, domain: string): Promise<void> {
  await withLock("hosts", async () => {
    const entries = (await getManagedDomains(hostsPath)).filter((d) => d !== domain);
    await writeManagedBlock(hostsPath, entries);
  });
}

// Rewrites the hosts file, stripping any existing managed block and appending a
// fresh one built from `entries`.
async function writeManagedBlock(hostsPath: string, entries: string[]): Promise<void> {
  const content = await readFile(hostsPath, "utf8");

  const kept: string[] = [];
  let inBlock = false;
  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    if (trimmed === BEGIN_MARKER) {
      inBlock = true;
      continue;
    }
    if (trimmed === END_MARKER) {
      inBlock = false;
      continue;
    }
    if (!inBlock) kept.push(line);
  }

  let result = kept.join("\n").replace(/\n+$/, "");
  if (entries.length > 0) {
    const block = [BEGIN_MARKER, ...entries.map((d) => `127.0.0.1\t${d}`), END_MARKER].join("\n");
    result = `${result}\n${block}\n`;
  } else {
    result = `${result}\n`;
  }

  writeFileAtomicSync(hostsPath, result);
}
